### PR TITLE
sql: Add Logical Column ID field to ColumnDescriptor

### PR DESCRIPTION
**Pull request by RichardJCai**
The LogicalColumnID field mimics the ColumnID field however LogicalColumnID may be swapped
between two columns whereas ColumnID cannot. LogicalColumnID is referenced for virtual tables
(pg_catalog, information_schema) and most notably affects column ordering for SHOW COLUMNS.

This LogicalColumnID field support swapping the order of two columns - currently only used for
ALTER COLUMN TYPE when a shadow column is created and swapped with it's original column.

Does not affect existing behaviour.

Release note: None

### CODE REVIEW

**Comment by rohany** [comment]
Can you update this PR to include the representation upgrade path on descriptor read?

**Comment by RichardJCai** [comment]
Will do

File: pkg/sql/sqlbase/structured.go [hunk 1]
```diff
@@ -4095,3 +4096,13 @@ func GenerateUniqueConstraintName(prefix string, nameExistsFunc func(name string
 	}
 	return name
 }
+
+// GetLogicalColumnID returns the LogicalColumnID if it has one, otherwise
+// returns the ID. The regular ColumnID is used for backwards compatibility.
+func (desc *ColumnDescriptor) GetLogicalColumnID() ColumnID {
```
Comment (rohany) [inline pkg/sql/sqlbase/structured.go:4102]: This function will basically be created by proto already -- i think you want to to populate this when you read the descriptor from disk

File: pkg/sql/sqlbase/structured.go [hunk 2]
```diff
@@ -886,15 +885,21 @@ func generatedFamilyName(familyID FamilyID, columnNames []string) string {
 // in a similar way.
 func (desc *TableDescriptor) MaybeFillInDescriptor(
 	ctx context.Context, protoGetter protoGetter,
-) error {
+) (bool, error) {
```
Comment (rohany) [inline pkg/sql/sqlbase/structured.go:888]: I don't think you want to change the signature of this function -- its return value is only used in one place, and that place is specifically meant for whether the FK representations changed. That function is a special case already, so I would be fine to not use MaybeFillDescriptor there.

**Review by rohany — COMMENTED** [review]
Can you also add a test that writes a descriptor to disk without logical column IDs set, and ensures that when we read the descriptor back we get the logical col IDs populated?

**Comment by RichardJCai** [comment]
> Can you also add a test that writes a descriptor to disk without logical column IDs set, and ensures that when we read the descriptor back we get the logical col IDs populated?

Created this test in logical_column_id_test.go

**Comment by rohany** [comment]
LGTM barring a review from @jordanlewis and comments on the new field on the column descriptor. I'd like to see:
* A description of the new field
* When it should be used/changed and what is the difference from column ID
* A note that this doesn't exist on old descriptors

**Comment by RichardJCai** [comment]
> LGTM barring a review from @jordanlewis and comments on the new field on the column descriptor. I'd like to see:
> 
> * A description of the new field
> * When it should be used/changed and what is the difference from column ID
> * A note that this doesn't exist on old descriptors

Added these.

**Comment by rohany** [comment]
comment looks good, ty

**Comment by rohany** [comment]
Just thought of this -- allocate IDs should set this for all new columns

**Comment by RichardJCai** [comment]
> Just thought of this -- allocate IDs should set this for all new columns

It currently does do this - AllocateIDs calls MaybeFillColumnIDs which fills out the ID and LogicalColumnID for new cols.

**Review by jordanlewis — COMMENTED** [review]
Hmm... I don't think that this will work in all cases yet.

Isn't it the case that things will go wrong if you have a table with columns (a, b, c), and you re-type b to a new type, so it gets logical id 2 but physical id 4? What happens in that case if you do `SELECT * FROM t` or `INSERT INTO t VALUES(1,2,3)`? Can you write a test for this and see what happens?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46992#-:-M4zhNku0lJA1Bk0OynR:bz4g3by)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @RichardJCai)

___
*[pkg/sql/sqlbase/structured.proto, line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46992#-M4zfrVs3bdmAtTcFw5h:-M4zfrVs3bdmAtTcFw5i:bmz9efs) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afcfd0decfe6aa47a932a723022b4a5ad80517f/pkg/sql/sqlbase/structured.proto#L132)):*
> ```protobuf
>   // computed column.
>   optional string compute_expr = 11;
>   // LogicalColumnID is used to order columns visually. This is because
> ```

It's not just visually. It's also for the purposes of SELECT * and INSERT INTO without column specifiers, right?

**Comment by rohany** [comment]
> It's not just visually. It's also for the purposes of SELECT * and INSERT INTO without column specifiers, right?

I think it's just for display in virtual tables -- we can safely use the column ordinal position in the other cases that you are thinking about (like selects or inserts).

**Comment by RichardJCai** [comment]
> Hmm... I don't think that this will work in all cases yet.
> 
> Isn't it the case that things will go wrong if you have a table with columns (a, b, c), and you re-type b to a new type, so it gets logical id 2 but physical id 4? What happens in that case if you do `SELECT * FROM t` or `INSERT INTO t VALUES(1,2,3)`? Can you write a test for this and see what happens?
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46992#-:-M4zhNku0lJA1Bk0OynR:bz4g3by)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @RichardJCai)
> 
> _[pkg/sql/sqlbase/structured.proto, line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46992#-M4zfrVs3bdmAtTcFw5h:-M4zfrVs3bdmAtTcFw5i:bmz9efs) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afcfd0decfe6aa47a932a723022b4a5ad80517f/pkg/sql/sqlbase/structured.proto#L132)):_
> 
> > ```protobuf
> >   // computed column.
> >   optional string compute_expr = 11;
> >   // LogicalColumnID is used to order columns visually. This is because
> > ```
> 
> It's not just visually. It's also for the purposes of SELECT * and INSERT INTO without column specifiers, right?

For the 1st point - I have a test for that in the WIP PR - it's working as intended.
[Here's the test
](https://github.com/cockroachdb/cockroach/pull/46933/files#diff-61f64739efcfb8f306a9a9001d2cfeb2R207)

For the 2nd point
No it doesn't affect the SELECT * and INSERT INTO orders - those rely on the ordering of the ColumnDescriptors in the TableDescriptor's columns slice.

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46992)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @RichardJCai)

___
*[pkg/sql/sqlbase/structured.go, line 1215 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46992#-M4zmSWjAHCbVFEuPSHE:-M4zmSWjAHCbVFEuPSHF:b6ruc4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afcfd0decfe6aa47a932a723022b4a5ad80517f/pkg/sql/sqlbase/structured.go#L1215)):*
> ```Go
> // if the ColumnID is set and the LogicalColumnID is not
> // it assigns the ColumnID as the LogicalColumnID.
> func (desc *TableDescriptor) MaybeFillInLogicalColumnID() {
> ```

Rather than have this method, would it be better to only populate these lazily? You could have a method called `GetLogicalColumnID` that either returns the ID or the LogicalColumnID if it is set.
___
*[pkg/sql/sqlbase/structured.proto, line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46992#-M4zfrVs3bdmAtTcFw5h:-M4zkC6-8wNHtNt1DLkV:b-qj9as8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afcfd0decfe6aa47a932a723022b4a5ad80517f/pkg/sql/sqlbase/structured.proto#L132)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

For the 1st point - I have a test for that in the WIP PR - it's working as intended. [Here's the test ](https://github.com/cockroachdb/cockroach/pull/46933/files#diff-61f64739efcfb8f306a9a9001d2cfeb2R207)

For the 2nd point No it doesn't affect the SELECT * and INSERT INTO orders - those rely on the ordering of the ColumnDescriptors in the TableDescriptor's columns slice. 
</blockquote></details>

Ah okay. That test belongs in this PR then.  Also, I would explain in this comment why having this ordering is necessary, on top of the ordering provided by the ColumnDescriptor order in the slice. It's too easy to forget.
___
*[pkg/sql/sqlbase/system.go, line 705 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46992#-M4zmpxZDkrEJxcW8KM1:-M4zmpxZDkrEJxcW8KM2:b-ifgw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afcfd0decfe6aa47a932a723022b4a5ad80517f/pkg/sql/sqlbase/system.go#L705)):*
> ```Go
> 		Version:                 1,
> 		Columns: []ColumnDescriptor{
> 			{Name: "key", ID: 1, LogicalColumnID: 1, Type: *types.String},
> ```

This sort of thing is why I'm not super crazy about requiring LogicalColumnID. I feel like the lazily populated method might avoid this pain.

**Comment by RichardJCai** [comment]
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46992)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @RichardJCai)
> 
> _[pkg/sql/sqlbase/structured.go, line 1215 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46992#-M4zmSWjAHCbVFEuPSHE:-M4zmSWjAHCbVFEuPSHF:b6ruc4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afcfd0decfe6aa47a932a723022b4a5ad80517f/pkg/sql/sqlbase/structured.go#L1215)):_
> 
> > ```go
> > // if the ColumnID is set and the LogicalColumnID is not
> > // it assigns the ColumnID as the LogicalColumnID.
> > func (desc *TableDescriptor) MaybeFillInLogicalColumnID() {
> > ```
> 
> Rather than have this method, would it be better to only populate these lazily? You could have a method called `GetLogicalColumnID` that either returns the ID or the LogicalColumnID if it is set.
> 
> _[pkg/sql/sqlbase/structured.proto, line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46992#-M4zfrVs3bdmAtTcFw5h:-M4zkC6-8wNHtNt1DLkV:b-qj9as8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afcfd0decfe6aa47a932a723022b4a5ad80517f/pkg/sql/sqlbase/structured.proto#L132)):_
> 
> Previously, RichardJCai (Richard Cai) wrote…
> Ah okay. That test belongs in this PR then. Also, I would explain in this comment why having this ordering is necessary, on top of the ordering provided by the ColumnDescriptor order in the slice. It's too easy to forget.
> 
> _[pkg/sql/sqlbase/system.go, line 705 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46992#-M4zmpxZDkrEJxcW8KM1:-M4zmpxZDkrEJxcW8KM2:b-ifgw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7afcfd0decfe6aa47a932a723022b4a5ad80517f/pkg/sql/sqlbase/system.go#L705)):_
> 
> > ```go
> > 		Version:                 1,
> > 		Columns: []ColumnDescriptor{
> > 			{Name: "key", ID: 1, LogicalColumnID: 1, Type: *types.String},
> > ```
> 
> This sort of thing is why I'm not super crazy about requiring LogicalColumnID. I feel like the lazily populated method might avoid this pain.

Haha I actually used the lazy method (exactly how you described) in the initial PR.

Regarding having the test in this PR - that logic test depends on the ALTER COLUMN TYPE feature which doesn't exist in this PR. I can't think of an easy way to test the case you described without ALTER COLUMN TYPE right now.

File: pkg/ccl/backupccl/restore_planning.go [hunk 3]
```diff
@@ -389,6 +389,8 @@ func allocateTableRewrites(
 // backup descriptors provided. if skipFKsWithNoMatchingTable is set, FKs whose
 // "other" table is missing from the set provided are omitted during the
 // upgrade, instead of causing an error to be returned.
+// Also assigns LogicalColumnIDs for TableDescriptors that do not have it,
```
Comment (rohany) [inline pkg/ccl/backupccl/restore_planning.go:392]: Comment is out of date

File: pkg/sql/sqlbase/structured.go [hunk 4]
```diff
@@ -387,6 +387,7 @@ func GetTableDescFromIDWithFKsChanged(
 	if err != nil {
 		return nil, false, err
 	}
+
```
Comment (rohany) [inline pkg/sql/sqlbase/structured.go:390]: unnecessary changes

File: pkg/sql/sqlbase/structured.go [hunk 5]
```diff
@@ -1198,6 +1201,7 @@ func (desc *MutableTableDescriptor) MaybeFillColumnID(
 	}
 	columnNames[c.Name] = columnID
 	c.ID = columnID
+	c.LogicalColumnID = columnID
```
Comment (rohany) [inline pkg/sql/sqlbase/structured.go:1204]: Should this still be done?

File: pkg/sql/sqlbase/structured.proto [hunk 6]
```diff
@@ -129,6 +129,19 @@ message ColumnDescriptor {
   // Expression to use to compute the value of this column if this is a
   // computed column.
   optional string compute_expr = 11;
+  // LogicalColumnID is lazily set, do not access directly,
```
Comment (rohany) [inline pkg/sql/sqlbase/structured.proto:132]: [nit]: Say something like "LogicalColumnID must be accessed through the accessor ... it is incorrect to access it directly.

File: pkg/sql/sqlbase/structured.proto [hunk 7]
```diff
@@ -129,6 +129,19 @@ message ColumnDescriptor {
...
+  // access using GetLogicalColumnID().
+  // LogicalColumnID is used to order columns visually. This is because
```
Comment (rohany) [inline pkg/sql/sqlbase/structured.proto:134]: I think the example here could be shrunk to just mention ordering in catalog tables, rather than the specific case.

**Review by rohany — APPROVED** [review]
LGTM with comment nits.

File: pkg/sql/sqlbase/structured.go (hunk 5 above)
Comment (RichardJCai) [reply pkg/sql/sqlbase/structured.go:1204]: I think this should still be done, this assigns the column a LogicalColumnID if it's created in an updated version. It makes sense to set it on creation for updated TableDescriptors have this and lazy load/set the LogicalColumnID from previous versions imo.

Comment (rohany) [reply pkg/sql/sqlbase/structured.go:1204]: Doing this kind of feels opposite from "only set logical column id if you are swapping columns" or something

Comment (RichardJCai) [reply pkg/sql/sqlbase/structured.go:1204]: I don't think "only set logical column id if you are swapping columns" is the right mantra to follow necessarily. That's the only use case right now but not necessarily in the future.

But I don't think LogicalColumnID is specific to swapping columns.

Comment (rohany) [reply pkg/sql/sqlbase/structured.go:1204]: sounds good

**Review by rohany — APPROVED** [review]

**Comment by rohany** [comment]
have you run make again? its failing because generated proto is different

**Comment by RichardJCai** [comment]
> have you run make again? its failing because generated proto is different

Yeah, the proto binary was different cause I added a comment

**Comment by RichardJCai** [comment]
bors r=rohany
